### PR TITLE
[Private Key] Add support for PuTTY private key file format (V3 and V2)

### LINKED ISSUE #341: Add support for PuTTY key files
One of the connections we want to use SSH.NET for requires us to use a PuTTY private key file, however that isn't currently supported by the SSH.NET library. Would it be possible to add support for a PuTTY key file to the ``Renci.SshNet.PrivateKeyFile`` class?

### LINKED ISSUE #431: Putty Keyfile Regex
In Renci.SshNet.PrivateKeyFile i added for Putty an Regex to test it, but putty is missing the encryption and salt, so i could test it friday, but maybe someone could give it a look.

[0001-Testadding-of-Putty-Regex.zip](https://github.com/sshnet/SSH.NET/files/1987927/0001-Testadding-of-Putty-Regex.zip)

Regards
Lord_Pinhead

### LINKED ISSUE #434: Getting "invalid private key file." error
Hi,

I am new to this SFTP communication. I have received the Private_Keys.ppk file from my IT team, which is as follows,
PuTTY-User-Key-File-2: ssh-rsa
Encryption: aes256-cbc
Comment: imported-openssh-key
Public-Lines: 6
AAAAB3NzaC1yc2EAAAADAQABAAABAQCXVRQ8E0LnZ1aFepN0NaLaavip8cO+EJSp
29PBOB2pUCkua/3IXO7q3coziwdJOhPElL6vh3dHDYtuEqrXPhkEli9mruf20ZVj
vI5k4VtlQh5hzjwfgujmLpSWCA0OiH+i8f7y3J3e0PqIDD1E15ZShLWCpzB9YuXM
+Qq2eXs47QMJ9vUYlB6ZcMMftPBp7nQMDni+Z2IHFp4yq1XXo8wRu9pyoKG+j5T9
Vbb1voUCPwgLlcXETt3J3pzS/P+/lake7eTJytJXfhMCn2DovGBxnvHywILLPNaz
HE5FVMxc/ZI8IXqpbGKvY0CRqVNNlZWLRVntjL4IOuDZYMqI1cUJ
Private-Lines: 14
vXQMpVg0cnoA8nkceo5oKf86/QcWFON+Gik1gCPBkus1V60Zd2LVMAHHN0WV8836
OpZzZTUyQrCa3sM/vz9mvQwx1FyZaDNLl6goXK4acneRNHMwblobSH0AGMjCc2x/
WIPtI4gjrwZF8TTNt4eo1kxnW/4E4zilI9U6Kv3h7QTYkvlyA3G0v33y85FQwmfz
maaBi/+Deayp6eGzyEASUqedb6DNVBel7KRs6D2sazywXZZs42zAVxTd+2g8B6D6
cU7a6GXSVqmxMIicYvCT/i6O5cJ+srPLrDK3yqAsOfye5SwgAq8lK3c5q/mEStlG
MHKZfskOXXoaTN9XEZX6w3fxKdMHd4tSSzub4EcpHGfjIGzc55ns5WRQWNa9QJ4o
vb0ImTfhODgU0OHYy6d0kio/u2BTayC543S37cojdBw76eTJOTWtqjoLDr90vrfE
+oMJ26gUOJKvI7Rjy/n5bBHORwTVLyeVKKRxIqlvUqgNvRQ13etZnoGRqXmYOC2t
+3v0CdM6JKB7A8OSyl9u1gjkTdj/eSOZExYzfrIJqlSEhsLH5XFbV8dz3SQW40mw
BDzetChkuQ5lzW556pZwCXuOAfuOGclJI/QvPdVc1CveMa8B3j2QM7SzIf0iiGX5
8MufxkWUf2ReVKv3Onh0xdIyp0qizkztvaIV2vP5f71ZnWDAfpnALmi/HdFfvPi0
9Y9nd8z0O0HYjmCphXydgoV8nJH13zmkXq2lVqlsoeTFLpHm+fuYKGG7U5ap52lT
eVvH6jmgLcyPnqVqpM23ZhBX+yxxQZR3NofkaCNpQOEChbMayr0qSbB1DwwRn11R
D4N1ywXDmxKjqxWvbhfHVUCY5C+v1irpSZwmFNAGhSIIl6MYVWMsX2dfSTOSGalT
Private-MAC: d11ad00691481e1f6f4bedb11621574df1c93566

While creating PrivateKeyFile object,
privateKeyFile = new PrivateKeyFile(ppkfilepath);

I am getting following error,
Message: invalid private key file
Stack Trace:    at Renci.SshNet.PrivateKeyFile.Open(Stream privateKey, String passPhrase)
   at Renci.SshNet.PrivateKeyFile..ctor(String fileName, String passPhrase)
   at Renci.SshNet.PrivateKeyFile..ctor(String fileName)
   at TestApp.sftpProtocol.connectSftpServer(String strHost, Int32 iPort, String strUserName, String strPassword, Int32 iTimeout, String PPKFileName) in D:\Project\SFTP Demo\demo-master\sftp_demo\SFTP\TestApp\sftpProtocol.cs:line 36

Please advice.

Note: I am able to connect to the sftp using winscp.

### LINKED ISSUE #510: is it possible to connect SFTP using PUTTY KEY string?
I am trying to connect SFTP using with PUTTY Key string through C# but could not able to connect to the server. received an exception while connecting to SFTP and the exception is **"Invalid Private key file".**
Please let me know with possible solution about using with SSH.NET.

### PR DESCRIPTION
**Pull request by scott-xu**
This PR adds support for PuTTY private key file format (V3 and V2) described in https://tartarus.org/~simon/putty-snapshots/htmldoc/AppendixC.html

Close https://github.com/sshnet/SSH.NET/issues/341
Close https://github.com/sshnet/SSH.NET/issues/431
Close https://github.com/sshnet/SSH.NET/issues/434
Close https://github.com/sshnet/SSH.NET/issues/510

**Notes**:
1. PuTTY private key file format V1 is not included because
   > PPK version 1 was a badly designed format, only used during initial development, and not recommended for production use.
2. The Argon2 algorithm is from latest BouncyCastle beta version. ~~It may subject to change, then we need to change accordingly. (Hope not). Also hope BouncyCastle makes a formal release before SSH.NET next release.~~ Update: BouncyCastle 2.5.0 is released.

### CODE REVIEW

File: src/Renci.SshNet/PrivateKeyFile.PuTTY.cs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+                                var keyData = Argon2(
+                                    _argon2Type,
+                                    Convert.ToInt32(_argon2Iterations),
+                                    Convert.ToInt32(_argon2Memory),
+                                    Convert.ToInt32(_argon2Parallelism),
+#if NET
+                                    Convert.FromHexString(_argon2Salt),
+#else
+                                    Org.BouncyCastle.Utilities.Encoders.Hex.Decode(_argon2Salt),
+#endif
```
Comment (scott-xu) [inline src/Renci.SshNet/PrivateKeyFile.PuTTY.cs:90]: We can change it to be the same as below code. Or change the below code to be the same as here.
https://github.com/sshnet/SSH.NET/blob/3ec45e1af31c84a184c518741d0542321d6deedf/src/Renci.SshNet/PrivateKeyFile.PKCS1.cs#L46-L50

File: src/Renci.SshNet/PrivateKeyFile.PuTTY.cs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                Debug.Assert(keyType == _algorithmName, $"{nameof(keyType)} is not the same with {nameof(_algorithmName)}");
+
+                var privateKeyReader = new SshDataReader(privateKey);
+
+                Key parsedKey;
+
+                switch (keyType)
+                {
+                    case "ssh-ed25519":
+                        parsedKey = new ED25519Key(privateKeyReader.ReadBignum2());
```
Comment (scott-xu) [inline src/Renci.SshNet/PrivateKeyFile.PuTTY.cs:177]: `ED25519Key` class does not have a constructor with **both** `publicKeyData` and `privateKeyData`. The current implementation is to generate public key from private key. Should we add a new constructor?
https://github.com/sshnet/SSH.NET/blob/3ec45e1af31c84a184c518741d0542321d6deedf/src/Renci.SshNet/Security/Cryptography/ED25519Key.cs#L98-L104

File: src/Renci.SshNet/PrivateKeyFile.PuTTY.cs (hunk 1 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/PrivateKeyFile.PuTTY.cs:90]: > Or change the below code to be the same as here.

sounds good

File: src/Renci.SshNet/PrivateKeyFile.PuTTY.cs (hunk 2 above)
Comment (Rob-Hague) [reply src/Renci.SshNet/PrivateKeyFile.PuTTY.cs:177]: it seems ok as it is, but I don't mind

File: src/Renci.SshNet/PrivateKeyFile.PuTTY.cs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+                var reference = Org.BouncyCastle.Utilities.Encoders.Hex.Decode(_mac);
+#endif
+                if (!macValue.SequenceEqual(reference))
+                {
+                    throw new SshException("MAC verification failed for PuTTY key file");
+                }
+
+                var publicKeyReader = new SshDataReader(_publicKey);
+                var keyType = publicKeyReader.ReadString(Encoding.UTF8);
+                Debug.Assert(keyType == _algorithmName, $"{nameof(keyType)} is not the same with {nameof(_algorithmName)}");
```
Comment (Rob-Hague) [inline src/Renci.SshNet/PrivateKeyFile.PuTTY.cs]:
```suggestion
                Debug.Assert(keyType == _algorithmName, $"{nameof(keyType)} is not the same as {nameof(_algorithmName)}");
```

**Review by Rob-Hague — COMMENTED** [review]
lgtm, will wait for a stable BC release

**Comment by Rob-Hague** [comment]
ooh a flaky test

**Review by Rob-Hague — APPROVED** [review]

**Pull request merged**